compiler/packages/babel-plugin-react-compiler/src/ReactiveScopes/PromoteUsedTemporaries.ts

```diff
@@ -148,6 +152,243 @@ class CollectPromotableTemporaries extends ReactiveFunctionVisitor<State> {
+        !this.#consts.has(identifier.id)
+      ) {
+        /*
+         * If the identifier hasn't been promoted but is marked as needing
+         * promotion by the logic in `visitInstruction`, and we've seen a
+         * use of it after said marking, promote it
+         */
+        promoteIdentifier(identifier, this.#promotable);
+      }
+    }
+  }
+
+  override visitInstruction(
+    instruction: ReactiveInstruction,
+    state: InterState,
+  ): void {
+    for (const lval of eachInstructionValueLValue(instruction.value)) {
+      CompilerError.invariant(lval.identifier.name != null, {
+        reason:
+          'PromoteInterposedTemporaries: Assignment targets not expected to be temporaries',
+        loc: instruction.loc,
+      });
+    }
+
+    switch (instruction.value.kind) {
+      case 'CallExpression':
+      case 'MethodCall':
+      case 'Await':
+      case 'PropertyStore':
+      case 'PropertyDelete':
+      case 'ComputedStore':
+      case 'ComputedDelete':
+      case 'PostfixUpdate':
+      case 'PrefixUpdate':
+      case 'StoreLocal':
+      case 'StoreContext':
+      case 'StoreGlobal':
+      case 'Destructure': {
+        /*
+         * Copy current state entries so that we can iterate over it
+         * later without hitting the entries added by visiting this
+         * instruction.
+         */
+        const preEntries = [...state.entries()];
+
+        let constStore = false;
+
+        if (
+          (instruction.value.kind === 'StoreContext' ||
+            instruction.value.kind === 'StoreLocal') &&
+          (instruction.value.lvalue.kind === 'Const' ||
+            instruction.value.lvalue.kind === 'HoistedConst')
+        ) {
+          /*
+           * If an identifier is const, we don't need to worry about it
+           * being mutated between being loaded and being used
+           */
+          this.#consts.add(instruction.value.lvalue.place.identifier.id);
+          constStore = true;
+        }
+        if (
+          instruction.value.kind === 'Destructure' &&
+          (instruction.value.lvalue.kind === 'Const' ||
+            instruction.value.lvalue.kind === 'HoistedConst')
+        ) {
+          [...eachPatternOperand(instruction.value.lvalue.pattern)].forEach(
+            ident => this.#consts.add(ident.identifier.id),
+          );
+          constStore = true;
+        }
+        if (instruction.value.kind === 'MethodCall') {
+          // Treat property of method call as constlike so we don't promote it.
+          this.#consts.add(instruction.value.property.identifier.id);
+        }
+
+        super.visitInstruction(instruction, state);
+
+        if (instruction.lvalue && instruction.lvalue.identifier.name === null) {
+          // Add this instruction's lvalue to the state, initially not marked as needing promotion
+          state.set(instruction.lvalue.identifier.id, [
+            instruction.lvalue.identifier,
+            false,
+          ]);
```

Comment:
seems like you could avoid creating `preEntries` (which will addd up since we create it for every such instruction) by deferring the set here until after you've iterated state